# Global AI Job Market & Salary Trends 2025  
## 🎯 Project Report – Targeted Analysis for SME Tech Recruiters

---
## Phase 1 – Understand the question and identify analysis questions
## 1.1. Project Context

In a rapidly evolving technological landscape, small and medium-sized enterprises (SMEs) in the **technology sector** face increased challenges in attracting and retaining AI talent. With rising competition, evolving skill demands, and salary inflation driven by innovation waves, **HR teams must understand both market dynamics and salary benchmarks** to make informed recruitment decisions.

This project is designed to **support SME tech recruiters** in optimizing their hiring strategy for AI professionals. Using a global dataset of over 15,000 AI-related job postings from 50+ countries, we provide market-level insights as well as a predictive salary model.

---

## 1.2. Project Objectives

The project consists of two main parts:

### 1.2.1 Descriptive Analysis
- Identify which **industries and job roles** dominate the current AI job market.
- Examine how **remote vs hybrid vs onsite work models** are evolving.
- Explore trending roles with respect to **compensation**, **benefits**, and **required skills**.
- Compare salary distributions and trends to understand market volatility.

### 1.2.2 Predictive Modeling
- Build a **salary prediction model** using variables such as skills, company size, job location, education, experience level, year, and job type.
- Quantify the **most influential factors** affecting compensation and explain their underlying patterns.

---

## 1.3. Analytical Questions

This study seeks to answer the following key questions:

1. **Which AI-related job titles are in highest demand across different quarters and months?**
2. **Which job roles benefit or suffer from AI disruption**, as seen through salary and demand metrics?
3. **What are the most essential skills** across industries and job roles, and how are they correlated with salary and experience?
4. **Which features most strongly correlate with salary**, including role, experience, skills, industry, and location?
5. **Are salary distributions within the same industry significantly different across job titles**, and what explains those differences?

---

## 1.4. Target Audience

This report is written for:
- **Hiring managers and recruiters** in small-to-medium tech companies
- **Workforce planners** seeking to adapt to AI-related labor trends
- **Data enthusiasts** and **junior analysts** looking to explore real-world HR analytics
- **Career developers** aiming to understand which AI skills offer the most market value


## Phase 2. Data Overview

This section provides an initial inspection of the dataset structure and key variables. The dataset contains over 15,000 AI job listings collected globally, each represented as a row with information about job title, salary, required skills, company characteristics, and metadata related to posting time and remote work ratio.

### Preview of the First 5 Records
We begin by displaying the first five rows to get an idea of the data format and content.

In [1]:
# Step 1: Load dataset and inspect the first 5 rows
# Purpose: To review column names, data types, and spot potential irregularities (e.g., missing values, formatting issues)
import pandas as pd
import warnings
warnings.filterwarnings('ignore')

# Read Dataset
df = pd.read_csv("C:/Users/theha/OneDrive/Desktop/Projects/Global AI Job Market & Salary Trends 2025/Dataset/raw_data_ai_job_dataset.csv")

# Screening of data structure
print(df.head())

    job_id              job_title  salary_usd salary_currency  \
0  AI00001  ai research scientist     90376.0             USD   
1  AI00002   AI Software Engineer         NaN             USD   
2  AI00003          AI SPECIALIST    152626.0             USD   
3  AI00004           NLP Enginner     80215.0             USD   
4  AI00005          AI Consultant     54624.0             EUR   

  experience_level employment_type company_location company_size  \
0               SE              CT            China            M   
1               EN              CT           Canada            M   
2               MI              FL      Switzerland            L   
3               SE              FL            India            M   
4               EN              PT           France            S   

  employee_residence  remote_ratio  \
0              China            50   
1            Ireland           100   
2        South Korea             0   
3              India            50   
4         

### Initial Observations

- The dataset includes a rich set of variables, such as:
  - `job_title`, `required_skills`, `education_required`, and `years_experience` for job qualifications
  - `salary_usd`, `salary_currency`, `company_size`, and `remote_ratio` for compensation and work structure
  - `company_location` and `employee_residence` for geographic scope
  - Time-based features such as `posting_date` and `application_deadline`

- Some salary values are missing (e.g., row 2), which may require imputation or exclusion.

- The `required_skills` column contains comma-separated values, which will later be transformed for skill-level analysis.

- Categorical columns are encoded in abbreviated forms:
  - `experience_level`: EN = Entry, MI = Mid, SE = Senior, EX = Executive
  - `employment_type`: FT, PT, CT, FL
  - `company_size`: S = Small, M = Medium, L = Large

These observations will guide the cleaning and transformation steps that follow.

### 2.1 Data Structure and Missing Values

To better understand the dataset structure and data quality, we examine the column types, non-null counts, and memory usage.

The dataset contains **15,150 rows** and **19 columns**, representing various aspects of AI-related job listings across different companies and regions. Below is a breakdown:

- **Numerical columns**: `salary_usd`, `salary_local`, `years_experience`, `job_description_length`, `benefits_score`, `remote_ratio`
- **Categorical/textual columns**: `job_title`, `company_location`, `experience_level`, `education_required`, `industry`, `employment_type`, etc.
- **Date columns**: `posting_date`, `application_deadline` (to be converted to datetime format)

### Missing Data Overview

Several columns contain missing values:
- `salary_usd`: 762 missing values
- `required_skills`: 755 missing
- `education_required`: 537 missing

These missing values will be addressed using appropriate imputation techniques or filtered out, depending on their role in the modeling or analysis phase.

In [2]:
# Step 2: Explore dataset structure and identify missing values
# Purpose:
# - Understand column types and non-null counts
# - Check for missing values that may impact analysis
# - Identify which columns need type conversion (e.g., date fields)

df.info()
df.isnull().sum().sort_values(ascending=False)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15150 entries, 0 to 15149
Data columns (total 19 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   job_id                  15150 non-null  object 
 1   job_title               15150 non-null  object 
 2   salary_usd              14398 non-null  float64
 3   salary_currency         15150 non-null  object 
 4   experience_level        15150 non-null  object 
 5   employment_type         15150 non-null  object 
 6   company_location        15150 non-null  object 
 7   company_size            15150 non-null  object 
 8   employee_residence      15150 non-null  object 
 9   remote_ratio            15150 non-null  int64  
 10  required_skills         14395 non-null  object 
 11  education_required      14613 non-null  object 
 12  years_experience        15150 non-null  int64  
 13  industry                15150 non-null  object 
 14  posting_date            15150 non-null

required_skills           755
salary_usd                752
education_required        537
job_title                   0
job_id                      0
employment_type             0
salary_currency             0
company_location            0
company_size                0
employee_residence          0
experience_level            0
remote_ratio                0
years_experience            0
industry                    0
posting_date                0
application_deadline        0
job_description_length      0
benefits_score              0
company_name                0
dtype: int64

### 2.2 Data Cleaning – Handling Duplicates

To ensure data integrity and avoid biased analysis, we begin the cleaning process by checking for duplicate job listings.

#### Duplicate Detection:
Using the `df.duplicated()` function, we identified several rows with identical values across all columns. These rows may represent:
- Redundant entries due to scraping processes
- Re-posted jobs without meaningful changes
- Manual entry errors

#### Action Taken:
We removed the duplicate rows using `df.drop_duplicates()` to preserve only unique job records. This step helps prevent overestimation in frequency-based analysis and improves model generalization in the predictive phase.

In [3]:
# Step 3: Identify and remove duplicate records
# Purpose:
# - Prevent bias in frequency counts and model training
# - Ensure each job listing is unique

# Count duplicates
duplicates = df[df.duplicated()]
print(f"Number of duplicate rows: {len(duplicates)}")

# Drop duplicates
df = df.drop_duplicates()
print(f"New dataset shape after removing duplicates: {df.shape}")

Number of duplicate rows: 140
New dataset shape after removing duplicates: (15010, 19)


After removing the duplicate rows, the dataset now contains **15010 rows**, ensuring that each job listing is unique and representative of the market.

### 2.3 Data Cleaning – Date Conversion and Temporal Features

To enable time-based analysis, we converted `posting_date` and `application_deadline` to proper datetime formats.

#### Actions Taken:
- Used `pd.to_datetime()` to standardize date formats and handle errors gracefully
- Created additional columns for:
  - `posting_month`: for monthly trend analysis
  - `posting_year`: for year-over-year comparison
  - `posting_quarter`: for seasonal/quarterly insights

These time features will be instrumental in tracking market evolution and modeling time-dependent trends.

In [4]:
# Step 2.3: Convert date fields to datetime format
# Purpose:
# - Enable temporal analysis (month, quarter, year)
# - Calculate job post duration and seasonal trends

df['posting_date'] = pd.to_datetime(df['posting_date'], errors='coerce')
df['application_deadline'] = pd.to_datetime(df['application_deadline'], errors='coerce')

# Create additional time-based features (optional but useful later)
df['posting_month'] = df['posting_date'].dt.month
df['posting_year'] = df['posting_date'].dt.year
df['posting_quarter'] = df['posting_date'].dt.to_period('Q')

### 2.4 Data Cleaning – Standardizing Job Titles

The `job_title` column contained many inconsistent and misspelled entries, which would affect grouping and aggregation for descriptive and predictive analyses.

#### Actions Taken:
- Converted all job titles to lowercase and stripped whitespace
- Mapped variants and typos (e.g., "ml ops engineer", "nlp specialist") into standardized job roles using a predefined dictionary
- Applied proper casing (e.g., "AI Software Engineer") using a smart formatting function

This ensures that each job role is uniquely and consistently represented, improving the accuracy of salary imputation and downstream modeling.

In [5]:
# Step 4: Standardize job titles
# Purpose:
# - Remove case/typo inconsistency in 'job_title'
# - Enable reliable grouping for salary and skill analysis

# Define job title mapping for normalization
job_mapping = {
    'nlp engineer': 'nlp engineer',
    'machine learning engineer': 'machine learning engineer',
    'robotics engineer': 'robotics engineer',
    'data scientist': 'data scientist',
    'autonomous systems engineer': 'autonomous systems engineer',
    'research scientist': 'research scientist',
    'data engineer': 'data engineer',
    'deep learning engineer': 'deep learning engineer',
    'ai consultant': 'ai consultant',
    'ai architect': 'ai architect',
    'ai software engineer': 'ai software engineer',
    'computer vision engineer': 'computer vision engineer',
    'principal data scientist': 'principal data scientist',
    'ai product manager': 'ai product manager'
    # Add more standard forms if needed
}

# Lowercase & map
df['job_title_lower'] = df['job_title'].str.strip().str.lower()
df['job_title_cleaned'] = df['job_title_lower'].map(job_mapping).fillna(df['job_title_lower'])

# Smart title formatting
def smart_title(job):
    job = job.title()
    job = job.replace('Ai', 'AI').replace('Nlp', 'NLP').replace('Ml', 'ML')
    return job

df['job_title_cleaned'] = df['job_title_cleaned'].apply(smart_title)

# Finalize column
df['job_title'] = df['job_title_cleaned']
df.drop(columns=['job_title_lower', 'job_title_cleaned'], inplace=True)

### 2.5 Handling Missing Values – Salary (salary_usd)

The `salary_usd` column had 743 missing values, which could negatively impact both descriptive statistics and model training.

#### Imputation Strategy:
- We grouped the data by `industry` and `job_title_cleaned` to capture typical salary ranges within each job function and domain.
- Then we filled the missing salaries with the **median** of each group.
- Median was chosen over mean due to the **right-skewed distribution** of salary (as observed in `df.describe()`), which is typical for compensation data.

This method balances contextual accuracy and robustness to outliers.

In [6]:
# Step 2.5: Impute missing values in salary_usd
# Strategy:
# - Group by industry and standardized job title
# - Fill missing salaries with the group median (robust to outliers)

df['salary_usd'] = df.groupby(['industry', 'job_title'])['salary_usd']\
                     .transform(lambda x: x.fillna(x.median()))

# Confirm all missing values handled
print(df['salary_usd'].isnull().sum())  # Expect 0

# Optional: Check salary distribution
print(df.describe())

1
          salary_usd  remote_ratio  years_experience  \
count   15009.000000  15010.000000      15010.000000   
mean   123826.313945     49.483678          7.196602   
min     32519.000000      0.000000          0.000000   
25%     72424.000000      0.000000          2.000000   
50%    100441.000000     50.000000          5.000000   
75%    145785.000000    100.000000         10.000000   
max    999999.000000    100.000000        100.000000   
std    106309.690532     40.811361         10.887406   

                        posting_date           application_deadline  \
count                          15010                          15010   
mean   2024-08-29 08:36:48.447701504  2024-10-11 21:33:01.558960896   
min              2024-01-01 00:00:00            2024-01-16 00:00:00   
25%              2024-04-29 00:00:00            2024-06-13 00:00:00   
50%              2024-08-28 00:00:00            2024-10-12 00:00:00   
75%              2024-12-29 00:00:00            2025-02-10 00:00:00

### 2.6 Handling Missing and Inconsistent Values – Education Requirement

The `education_required` column included inconsistent text formatting (e.g., "phd", " PhD ", "nan") and 533 missing values.

#### Cleaning Steps:
- Standardized text to lowercase and stripped whitespace
- Converted string `"nan"` values to actual `NaN`
- Unified capitalization (e.g., "Phd" → "PhD")

#### Imputation Strategy:
- Used the **mode of `education_required` per `job_title`** to fill missing values, assuming that education requirements are job-specific.
- Applied this logic using `.groupby()` and `.apply()` for row-wise imputation.

After these steps, the column is ready for categorical analysis and modeling.

In [7]:
# Step 2.6: Clean and impute missing values for 'education_required'
# Purpose:
# - Standardize inconsistent labels
# - Fill missing values based on typical education per job_title

import numpy as np

# 1. Normalize and clean formatting
df['education_required'] = df['education_required'].str.lower().str.strip()
df['education_required'] = df['education_required'].replace('nan', np.nan)

# 2. Impute missing values using mode by job_title
edu_mode_by_job = df.groupby('job_title')['education_required']\
                    .agg(lambda x: x.mode().iloc[0] if not x.mode().empty else np.nan)

def impute_education(row):
    if pd.isna(row['education_required']):
        return edu_mode_by_job.get(row['job_title'], 'unknown')
    else:
        return row['education_required']

df['education_required'] = df.apply(impute_education, axis=1)

# 3. Capitalize properly: "phd" → "PhD"
df['education_required'] = df['education_required'].str.title()
df['education_required'] = df['education_required'].replace({'Phd': 'PhD'})

### 2.7 Handling Missing Values – Required Skills

The `required_skills` column contains key information for job qualification analysis and model input. However, 750 entries were missing.

#### Strategy:
- **Step 1**: Convert skill strings into lists and standardize formatting
- **Step 2**: Group data by `job_title`, `industry`, and `education_required`, and identify the most common 5 skills within each group
- **Step 3**: Apply imputation using these skill sets for missing rows
- **Step 4**: For rows where the full group is unavailable, apply a fallback method using the most common skills for each `job_title`

After processing, all missing values in `required_skills` were successfully filled.

> Note: This preserves job-specific relevance while ensuring complete data for downstream analysis and modeling.

In [8]:
# Step 2.7: Impute missing values in 'required_skills'
# Purpose:
# - Fill missing skill sets based on most common combination for similar roles

from collections import Counter

# Step 1: Prepare data
df_skills = df[df['required_skills'].notna()].copy()
df_skills['required_skills_list'] = df_skills['required_skills'].str.strip().str.split(',')

# Step 2: Group by relevant dimensions and get most common skill sets
group_cols = ['job_title', 'industry', 'education_required']
grouped_skills = (
    df_skills.groupby(group_cols)['required_skills_list']
    .apply(lambda x: [skill.strip() for sublist in x for skill in sublist])
    .apply(lambda x: Counter(x).most_common(5))
)

# Convert to string for replacement
grouped_skills = grouped_skills.apply(lambda x: ', '.join([s[0] for s in x]))

# Step 3: Fill function
def fill_required_skills(row):
    if pd.isna(row['required_skills']):
        key = (row['job_title'], row['industry'], row['education_required'])
        return grouped_skills.get(key, np.nan)
    else:
        return row['required_skills']

df['required_skills_filled'] = df.apply(fill_required_skills, axis=1)

# Step 4: Backup method using only job_title
# Build fallback skill set by job_title
job_title_skills = {}
for title, group in df_skills.groupby('job_title'):
    all_skills = [skill.strip() for sublist in group['required_skills_list'] for skill in sublist]
    most_common = Counter(all_skills).most_common(5)
    job_title_skills[title] = ', '.join([s[0] for s in most_common])

def fallback_fill(row):
    if pd.isna(row['required_skills_filled']):
        return job_title_skills.get(row['job_title'], None)
    else:
        return row['required_skills_filled']

df['required_skills'] = df.apply(fallback_fill, axis=1)

# Final check
print("Missing values remaining in 'required_skills':", df['required_skills'].isna().sum())

Missing values remaining in 'required_skills': 0
